services/vision/obstacledepth/obstacle_depth.go

```diff
@@ -0,0 +1,377 @@
+// Package obstacledepth uses an underlying depth camera to fulfill GetObjectPointClouds,
+// using the method outlined in (Manduchi, Roberto, et al. "Obstacle detection and terrain classification
+// for autonomous off-road navigation." Autonomous robots 18 (2005): 81-102.)
+package obstacledepth
+
+import (
+	"context"
+	"image"
+	"math"
+	"strconv"
+	"sync"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"github.com/muesli/clusters"
+	"github.com/muesli/kmeans"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/pointcloud"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/rimage"
+	"go.viam.com/rdk/rimage/transform"
+	"go.viam.com/rdk/robot"
+	svision "go.viam.com/rdk/services/vision"
+	"go.viam.com/rdk/spatialmath"
+	"go.viam.com/rdk/utils"
+	vision "go.viam.com/rdk/vision"
+	"go.viam.com/rdk/vision/segmentation"
+)
+
+var model = resource.DefaultModelFamily.WithModel("obstacle_depth")
+
+// ObstaclesDepthConfig specifies the parameters to be used for the obstacle depth service.
+type ObstaclesDepthConfig struct {
+	K          int                                `json:"k"`
+	Hmin       float64                            `json:"hmin"`
+	Hmax       float64                            `json:"hmax"`
+	ThetaMax   float64                            `json:"theta_max"`
+	ReturnPCDs bool                               `json:"return_pcds"`
+	Intrinsics *transform.PinholeCameraIntrinsics `json:"intrinsic_parameters"`
+}
+
+// obsPoint is a useful intermediate struct (which pts are obstacles).
+type obsPoint struct {
```

Comment:
I think this struct might be less helpful than first imagined. intermediate struct can create a lot of unnecessary copies, and its best to find ways to eliminate them when possible